### ISSUE TITLE
How to deserialize a class with a relationship that is a derived class?

### ISSUE THREAD

**Issue by marioGomezDev** [json-api-dotnet/JsonApiDotNetCore#696]
## Description
My domain are set up like this:

```
abstact class Person {
	...
}

class Student {
	...
}

class Teacher{
	...
}

class Article {
	Person Author
}

```
When I post an Article, it fails to deserialize the specific Teacher or Student class, how can I fix it?
...

## Environment

- JsonApiDotNetCore Version: 4.0.0.0

**Comment by maurei** [comment]
Thanks for reporting this issue. There might be some issues with the (de)serialization of inherited properties, but to be sure what is going wrong and to see if we need to patch anything in the internals, we would need some more details. Preferably if you could open up a reproduction repository for us to step into and debug, or even better would be a PR with a few tests that reproduce the issue in the JsonApiExample project

**Comment by marioGomezDev** [comment]
In this repository, you can see the issue
[https://github.com/marioGomezDev/DeserializeIssue.git](url)

Test example:
[https://www.getpostman.com/collections/d8f6c4df7407152a9f0b](url)

**Comment by bart-degreed** [comment]
Hi @marioGomezDev, The repro repository your provided targets JADNC v3, so I [upgraded](https://github.com/marioGomezDev/DeserializeIssue/pull/1) it to .NET Core 3.1 and JADNC 4-beta1. After that, I was able to reproduce your issue using the next request:

```http
POST https://localhost:44308/articles HTTP/1.1

{
  "data": {
    "type": "articles",
    "attributes": {
      "article-prop": "AP-new"
    },
    "relationships": {
      "author": {
        "data": {
          "type": "students",
          "id": "1"
        }
      }
    }
  }
}
```

This crashes in the deserialization process, because it tries to create an instance of abstract Person class. Instead it should look at the actual type ("students" in this case) and so create an instance of Student. I have a preliminary fix at #828.

For context, the next request shows the sample data:

```http
GET https://localhost:44308/articles?include=author HTTP/1.1
```
```json
{
  "links": {
    "self": "https://localhost:44308/articles?include=author"
  },
  "data": [
    {
      "type": "articles",
      "id": "1",
      "attributes": {
        "article-prop": "ArticleProp 1"
      },
      "relationships": {
        "author": {
          "links": {
            "self": "https://localhost:44308/articles/1/relationships/author",
            "related": "https://localhost:44308/articles/1/author"
          },
          "data": {
            "type": "students",
            "id": "1"
          }
        }
      },
      "links": {
        "self": "https://localhost:44308/articles/1"
      }
    },
    {
      "type": "articles",
      "id": "2",
      "attributes": {
        "article-prop": "ArticleProp 2"
      },
      "relationships": {
        "author": {
          "links": {
            "self": "https://localhost:44308/articles/2/relationships/author",
            "related": "https://localhost:44308/articles/2/author"
          },
          "data": {
            "type": "teachers",
            "id": "2"
          }
        }
      },
      "links": {
        "self": "https://localhost:44308/articles/2"
      }
    }
  ],
  "included": [
    {
      "type": "students",
      "id": "1",
      "attributes": {
        "student-prop": "StudentProp 1",
        "person-prop": "PersonProp Student 1"
      },
      "links": {
        "self": "https://localhost:44308/students/1"
      }
    },
    {
      "type": "teachers",
      "id": "2",
      "attributes": {
        "teacher-prop": "TeacherProp 2",
        "person-prop": "PersonProp Teacher 1"
      },
      "links": {
        "self": "https://localhost:44308/teachers/2"
      }
    }
  ]
}
```
